### PR TITLE
[Update][Proposal] Improve Canvas Documentation

### PR DESCRIPTION
**Pull request by MasahideMori-SimpleAppli**
This is an additional patch to the previously submitted patch.

Thanks to @flar for suggestions for making it better.

https://github.com/flutter/engine/pull/30069

- This patch improves the documentation for the Vertices.raw constructor.

Also, this time, when I Fetch upstream from the browser (flutter / engine), I got an Actions error saying "GitHub token env var is not set.", But I haven't solved it.
I'm not sure what this is because I don't use Actions,
Please let me know if you have any problems with this patch.
If the patch is okay, ignore my problem with Actions.
Thank you.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -3974,22 +3974,33 @@ class Vertices extends NativeFieldWrapperClass1 {
 
   /// Creates a set of vertex data for use with [Canvas.drawVertices], directly
   /// using the encoding methods of [new Vertices].
+  /// Note that this constructor uses a raw typed data list,
```
Comment (flar) [inline lib/ui/painting.dart:3977]: uses raw typed data lists,

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -3974,22 +3974,33 @@ class Vertices extends NativeFieldWrapperClass1 {
...
+  /// so it runs faster than Dart list version
```
Comment (flar) [inline lib/ui/painting.dart:3978]: faster than the [Vertices()] constructor

(Make sure it links correctly to the regular constructor)

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -3974,22 +3974,33 @@ class Vertices extends NativeFieldWrapperClass1 {
 
   /// Creates a set of vertex data for use with [Canvas.drawVertices], directly
   /// using the encoding methods of [new Vertices].
+  /// Note that this constructor uses a raw typed data list,
+  /// so it runs faster than Dart list version 
+  /// because it doesn't require any conversion from Dart list.
```
Comment (flar) [inline lib/ui/painting.dart:3979]: from Dart lists.

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -3974,22 +3974,33 @@ class Vertices extends NativeFieldWrapperClass1 {
...
+  /// The cut part is applied to the triangular mesh.
+  /// Note that the [textureCoordinates] are the coordinates on the image.
   ///
   /// The [colors] list is interpreted as a list of RGBA encoded colors, similar
   /// to [Color.value]. It must be half length of [positions] if it is not
   /// null.
   ///
   /// If the [indices] list is provided, all values in the list must be
   /// valid index values for [positions].
+  ///
```
Comment (flar) [inline lib/ui/painting.dart:4002]: No extra line here, you are continuing to describe the same parameter as the previous paragrph.

File: lib/ui/painting.dart [hunk 5]
```diff
@@ -3974,22 +3974,33 @@ class Vertices extends NativeFieldWrapperClass1 {
...
   /// The [mode] parameter must not be null.
   ///
   /// The [positions] list is interpreted as a list of repeated pairs of x,y
   /// coordinates. It must not be null.
+  /// The [positions] parameter is a list of triangular mesh vertices(xy).
   ///
   /// The [textureCoordinates] list is interpreted as a list of repeated pairs
   /// of x,y coordinates, and must be the same length of [positions] if it
   /// is not null.
+  ///
```
Comment (flar) [inline lib/ui/painting.dart:3990]: No extra line here, both of these paragraphs describe the same parameter.

File: lib/ui/painting.dart [hunk 6]
```diff
@@ -3974,22 +3974,33 @@ class Vertices extends NativeFieldWrapperClass1 {
...
   /// using the encoding methods of [new Vertices].
+  /// Note that this constructor uses a raw typed data list,
+  /// so it runs faster than Dart list version 
+  /// because it doesn't require any conversion from Dart list.
   ///
   /// The [mode] parameter must not be null.
   ///
   /// The [positions] list is interpreted as a list of repeated pairs of x,y
   /// coordinates. It must not be null.
+  /// The [positions] parameter is a list of triangular mesh vertices(xy).
```
Comment (flar) [inline lib/ui/painting.dart:3985]: These 2 sentences should be combined, such as

The [positions] parameter is a list of triangular mesh vertices and is interpreted as ...

Comment (flar) [reply lib/ui/painting.dart:3985]: (And no need to specify xy twice as the rest of the first sentence also mentions that they are x,y coordinates.)

**Review by flar — CHANGES_REQUESTED** [review]
There is another problem with the docs here in that the colors list says it is RGBA, but the correct format is ARGB. (which is the same format as Color.value)

One more issue which happens with the Canvas.drawVertices call that you also updated with the previous PR - it does not describe how the BlendMode parameter is used. We could say (in the documentation for drawVertices):

The [mode] parameter is used to control how the colors in the [vertices] are combined with the colors in the [paint]. If there are no colors specified in [vertices] then the [mode] has no effect. If there are colors in the [vertices], then the color taken from the [Shader] or [Color] in the [paint] is blended with the colors specified in the [vertices] using the [mode] parameter. For purposes of this blending, the colors from the [paint] are considered the source and the colors from the [vertices] are considered the destination.

**Comment by MasahideMori-SimpleAppli** [comment]
Hi @flar,
Thank you for your comment.
I fixed those issues.
Is the position of the description of drawVertices [blendMode] appropriate here?

Your comments were detailed and very helpful to me.
If there are still corrections left, I would appreciate it if you could point them out.
Thank you for your all help in advance.

**Review by flar — APPROVED** [review]
Looks Great! Thanks for doing this!

**Comment by MasahideMori-SimpleAppli** [comment]
Thank you, too!
I learned a lot.

**Comment by chinmaygarde** [comment]
The linter failure is real. There are couple of trailing whitespaces in this patch:

```
ERROR: Whitespace check failed. The following files have trailing spaces:
  /b/s/w/ir/cache/builder/src/flutter/lib/ui/painting.dart:3976:  /// so it runs faster than the [Vertices()] constructor 
  /b/s/w/ir/cache/builder/src/flutter/lib/ui/painting.dart:3981:  /// The [positions] parameter is a list of triangular mesh vertices and 
```

Let's address these nits and land ASAP. Thanks.

**Comment by MasahideMori-SimpleAppli** [comment]
Hi @chinmaygarde

Thank you for always helping me out!
I fixed the problem.

**Comment by flar** [comment]
I think I've seen that failure in the Linux Smoke Tests before. You might have to rebase this PR to the Top of Tree to fix it.

**Comment by MasahideMori-SimpleAppli** [comment]
Hi @flar,
Thank you for your advice!
I tried rebase. Thank you again.

**Review by zanderso — APPROVED** [review]
Lgtm

**Pull request merged**